### DISCUSSION TITLE
[beginners-guide] Vim 入門書に載せるコンテンツについて

### DISCUSSION THREAD

**Discussion by monaqa** [vim-jp-book/zenn-content-book#1] [General]
Vimの様々なコンテンツに関する入門書を企画しています。

- 概要
    - Vim のあらゆるトピックに関する入門書。
    - トピックの例
        - Vim の標準機能（オペレータや Quickfix）
        - プラグインの入れ方
        - プラグインの作り方
        - Vim へのコントリビュートの方法
- 想定読者
    - 初心者から上級者まで、幅広い読者層を想定しています。
        - Vim の存在は知っているが、基本的な操作方法を知らない、自信がない人
        - Vim の基本操作は知っているがプラグインの導入方法や使い方が知りたい人
        - Vim が好きで、Vim に貢献したい人

ここでは、具体的にどういったトピックを入れるべきか議論したいです。

**Comment by monaqa** [comment]
現時点でのトピック案

3部 (+ 1部) 構成に分ける想定。

1. Vim本来の操作
    - オペレータ
    - テキストオブジェクト
    - exコマンド
    - Quickfix
    - 正規表現
2. プラグインの活用入門
    - VSCodeからの移行
    - プラグインの探し方
    - プラグインのインストール手順
        - vim-plug
        - lazy.nvim
        - dpp.vim
3. コントリビュート入門
    - プラグイン作成
        - Vim script
        - Lua
        - Vim9script
        - denops.vim
    - コミュニティ入門
        - Vim-jp 入門
        - Vimrc読書会
        - イベント参加
    - Vim本体へのパッチ作成
4. おまけ
    - ブラウザ拡張(vimium)
    - イベント主催について

**Comment by monaqa** [comment]
一旦自由にトピックを募りたいと思います。
こんな感じのフォーマットで、ここにリプライする形で記載していきましょうか。
他の人と被っても一旦気にしないで書いてください。

```markdown
## 読みたい

- （自分が読んでみたいトピックを箇条書きで記入）

## 書きたい

- （自分が執筆したいトピックを箇条書きで記入）

## あったほうが良さそう

- （自分の興味や執筆願望とは別に、客観的にあったほうが良さそうなものを箇条書きで記入）
```

**Reply by monaqa** [reply]
## 読みたい

- quickfix list / location list 入門
- Vim へのコントリビュート入門
- Lua プラグイン作成入門

## 書きたい

- モーション入門
- ドットリピート入門
- 【追加】tree-sitter

## あったほうが良さそう

- vimrc 入門
    - vimrc の書き方、典型的なディレクトリ構成などの解説。
    - オプションの指定方法やキーマッピングなどもここに入る？

**Reply by staticWagomU** [reply]
## 読みたい

- 1章

## 書きたい

- プラグインの探し方
- プラグインのインストール手順
	- vim-plug

## あったほうが良さそう

- プラグイン作成に関わらないLua入門、Vim script入門

**Reply by Omochice** [reply]
## 読みたい

- quickfix
- dot repeat
- vim9script

## 書きたい

- ブラウザ拡張(vimium)

## あったほうが良さそう

- コードをどう読むか
  - ここtextobjでいけそうだなとかdot repeatできそうだなとか（思考術?）

**Reply by atusy** [reply]
## 読みたい

- Quickfix
- 正規表現

## 書きたい

- 1章から何か
    - 勉強も兼ねれそうなので……
        - 正規表現かなあ？

## あったほうが良さそう

- vimrcの書き方
    - 巷にあふれてるからいらない説もある……？
    - でもプラグイン活用術の手前に欲しい気がする
- オレオレVim利用術的なの
    - せっかく多数で共著するので色んな人の色を見たい
    - 起動しっぱなし派/すぐ閉じる派
    - tabpage使う派/使わない派
    - とか

**Reply by thinca** [reply]
## 読みたい

- Vim 本体のコードの読み方
- LSP (今の時代 Vim とは切り離せないと思うので Vim トピックに入れてよさそう)

## 書きたい

- Vim9 script (勉強するところから…)

## あったほうが良さそう

- Vim/Neovim の違い、選び方

**Reply by eihigh** [reply]
### 読みたい
* ex commands
* quickfix
* 本体へのパッチ

### 書きたい
* vscodeからの移行
* (lsp周りの) プラグインの探し方

### あった方が良さそう
特になし（これだけあれば大満足）

**Reply by uga-rosa** [reply]
## 読みたい

- Quickfix
- Vim本体へのパッチ作成

## 書きたい

- Lua
- denops.vim

## あったほうが良さそう

- 今のところは特になし

**Reply by NI57721** [reply]
## 読みたい

- Quickfix
- 正規表現
- Vim9 script

## 書きたい

- Quickfix
- 正規表現

## あったほうが良さそう

- textobject の活用術みたいなのは読んでみたいかも

**Reply by kuuote** [reply]
## 読みたい

- Vim 本体へのコントリビュート
- Vim を用いた編集の思考術
- プラグインの探し方

## 書きたい

- オペレータ
- テキストオブジェクト
- Language Server Protocol
- Vim 本体のコードの読み方

## あったほうが良さそう

- help の読み方とか？(help に書いてあるけど)
- atusy さんが出してるオレオレ Vim 利用術の案は気になる
- 最小構成入門

**Reply by yasunori0418** [reply]
## 読みたい

- プラグインの探し方
- Quickfix
- 正規表現
- Vim9 script
- プラグインの探し方
- Vim 本体のコードの読み方
- 

## 書きたい

- Vim を用いた編集の思考術
- オペレータ
- テキストオブジェクト


## あったほうが良さそう

- プラグインとは
- プラグインを読み込むとは
- vimtutorについて
- 最小構成入門

**Reply by sankantsu** [reply]
## 読みたい

- quickfix
- operator, textobject
- vim 利用術

## 書きたい

勉強がてら書いてみたいかなと思うもの
だいぶ被りますが...

- operator, textobject
- Language Server Protocol (Neovim なら書けそう, Vim 側はあまりわからないかも)
- プラグイン導入 (lazy.nvim)

## あったほうが良さそう

すでに出てるもの

- プラグインとは

思いつくままにいくつか

- autocmd
- key mapping の定義
- window, tab, terminal あたりの利用 
- treesitter

autocmd, key mapping あたりは、Vim 設定入門として .vimrc (あるいは init.lua) の紹介と一緒にやっても良いか

- プラグイン活用で、具体的にいくつか便利なプラグインを紹介する
  - いろいろ記事はあるのでそれへの参照でいいという説はある

## 構成について

Zenn book でやるなら、部ごとに本つくるって感じでしょうか?
確か chapter = 記事 を並べる程度の構造しかつくれなかったと思うので

**Reply by kyoh86** [reply]
## 読みたい

- Mark
- レジスタ
- Shada

## 書きたい

- :help の魅力
- （一番小さな設定の足がかりとしての）Exコマンド

## あったほうが良さそう

- VimとNeovim
- なぜ今Vimなのか

**Reply by satorunooshie** [reply]
## 読みたい

- 非同期処理
- terminalの良さ
- exコマンド
- Vim本体のコードの読み方
- 肥大化したVim Scriptのディレクトリ構成
- 潰していいキーマッピング

## 書きたい

- register
- マクロ

## あったほうが良さそう
- マクロ
- viminfo
- カラースキーマについて軽く？

**Reply by mityu** [reply]
## 読みたい

-  Vim 本来の操作/Quickfix
- コントリビュート入門/プラグイン作成/Lua
- Vim 本体のコードの読み方

## 書きたい

- Vim9 script
- Vim 本体のプラグイン機構
- ex コマンド

## あったほうが良さそう

- （プラグイン作成の項に含まれるかもしれませんが）Vim 本体のプラグイン機構。runtimepath や packpath、Vim 本体のプラグインの読み込み方などを想定しています。パッケージマネージャの裏側と言った方が良いのかもしれません。

**Reply by AmaiSaeta** [reply]
## 読みたい

- Quickfixとlocationリストについて
- LSPについて

## 書きたい

- 

## あったほうが良さそう

- 「（ノーマルモード等での）コマンド」と「Exコマンド」の違い（初心者向けとして。「コマンド」→「（キーボード）ショートカット」、「Exコマンド」→「コマンド」の誤解が発生しそうな気がする――というか私がそうだった――ので）
- 「Exコマンド」と「Vim script」って同じもの？違うもの？（コラム的にちょっと触れるだけで良いと思うけれど）
- プラグインマネージャを使わないプラグインの導入方法（これまたコラム的に。「現代ではプラグインマネージャを使う事を推奨しますが〜」の但し書き付きでサラッと）
- 「Vim使うと変なファイルが出来るんだけどナニコレ？（スワップファイル、バックアップファイルについて）」
- Vimってどういうソフト？Bram Moolenaarって誰？（プログラミング言語の入門書の最初に在るような奴）
- ウガンダ

**Comment by monaqa** [comment]
# トピックと担当者

一旦細かめに分けています。まとめられそうなトピックはまとめるかもしれません。

## Vim 基本機能入門

|トピック                                 |担当者候補                  |
|-----------------------------------------|----------------------------|
|Vimの概要と特徴（なぜ今Vimなのか）       |                            |
|Vim/Neovimの違い、選び方                 |                            |
|モーション                               |monaqa                      |
|オペレータ・テキストオブジェクト         |kuu, yasunori0418, sankantsu|
|ドットリピート                           |monaqa                      |
|mark                              |                            |
|register                             |satorunooshie|
|マクロ                             |satorunooshie|
|quickfix list, location list             |NI57721                     |
|Vim の正規表現                           |atusy?, NI57721             |
|Vim を用いた編集の思考術                 |yasunori0418                |
|help の読み方                            |kyoh86|
|buffer/window/tabpage (+ terminal window)|                            |
|swapfile|                            |

## Vim 設定・カスタマイズ入門

|トピック                                 |担当者候補                  |
|-----------------------------------------|----------------------------|
|Ex コマンド                              |kyoh86?, mityu|
|言語としての Vim script                  |                            |
|言語としての Lua                         |uga-rosa                    |
|.vimrc                                   |kyoh86?|
|autocmd                                  |                            |
|key mapping                              |                            |
|shada                              |                            |
|colorscheme|                       |

## Vim プラグイン活用入門

|トピック                                |担当者候補             |
|----------------------------------------|-----------------------|
|プラグインとは・プラグインを読み込むとは||
|Visual Studio Code からの移行           |eihigh                 |
|Language Server Protocol                |eihigh?, kuu, sankantsu|
|プラグインの探し方                      |staticWagomU, eihigh?  |
|プラグインの機構、仕組み|mityu|
|プラグインのインストール - packpath（本体に入っているもの）     |mityu           |
|プラグインのインストール - vim-plug     |staticWagomU           |
|プラグインのインストール - lazy.nvim     |sankantsu              |
|tree-sitter                              |monaqa                      |

## Vim コントリビュート入門

|トピック                    |担当者候補|
|----------------------------|----------|
|プラグイン作成 - Lua        |uga-rosa  |
|プラグイン作成 - denops     |uga-rosa  |
|プラグイン作成 - vim9script |thinca, mityu    |
|非同期処理 | |
|Vim 本体のコードの読み方    |kuu       |
|Vim 本体へのコントリビュート|          |
|最小構成入門                |          |

## おまけ？

|トピック       |担当者候補|
|---------------|----------|
|ブラウザ拡張   |Omochice  |
|Vim 利用術     |          |
|vimtutor の紹介|          |
|Vim に関する有用な情報入手先の紹介| |

※ 入門というより紹介になりそうなもの、Vim 本体から離れるものはおまけとして載せようかと思っています

## 別トピックにまとめられそうなもの

- .vimrc
    - 肥大化した Vim script のディレクトリ構成
- key mapping
    - 潰していいキーマッピング
- Ex コマンド
    - ノーマルモードコマンドと Ex コマンドの区別
    - Vim script との関連
- Vimの概要と特徴（なぜ今Vimなのか）
    - Bram 氏について
    - ウガンダ

**Comment by kuuote** [comment]
選択肢がどうしても増えそうな部分については絞るか、全部書いた上でおすすめの提案をするか決めた方がいいかもと思いました(例えばプラグインマネージャの導入とか)

**Reply by yasunori0418** [reply]
プラグインの一部としてLSPの紹介があるのは良さそうだけど、どこまで書くべきなのかvimの入門という視点で考えた時に、範囲を考慮しないと膨大になりそうな気がしてます。

**Reply by sankantsu** [reply]
LSP は高級エディタ的に使うには必須級だと思っているので、紹介はあると良さそうですね
Vim/Neovim の入門の一部として書くなら、下の内容ぐらいでしょうか
- LSP とは
- LSP でどんな恩恵が得られるか
- LSP 設定 (vim-lsp)
- LSP 設定 (nvim-lspconfig)
補完とかは欲しい気もするけど、結構複雑でもあるので微妙

**Comment by monaqa** [comment]
https://github.com/vim-jp-book/zenn-content-book/discussions/1#discussioncomment-8493296

> Zenn book でやるなら、部ごとに本つくるって感じでしょうか?
> 確か chapter = 記事 を並べる程度の構造しかつくれなかったと思うので

その手もありますし、[この本](https://zenn.dev/kazutxt/books/flutter_practice_introduction) みたいに無理やり載せてしまう手もあると思います。

**Reply by monaqa** [reply]
Zenn はご指摘どおり階層構造の表現力が高くないので元々「部 - トピック」ぐらいの構造に留めようと思っていたんですが、
思った以上に第1部のトピックが増えそうですね。
第1部となっているものを2つに分割してもいいかも。

1. Vim の概要と基本機能
    - Vim の特徴、Vim/Neovim の区別とか
    - オペレータとか mark とか register とか
    - help の読み方 とか

2. Vim の設定方法
    - key mapping とか autocmd とか
    - Vim script/Lua 入門とか

- Ex コマンド はどっちに入れるか微妙

**Reply by monaqa** [reply]
部構造をあえて設けずに、各章の対象読者やレベル感は最初の章で全て説明するという選択肢もありそうです。
そのほうが Zenn 的には自然、かつ全章が独立であるということが強調されて良いかもしれません。

**Comment by makizo** [comment]
## 読みたい

- colorschemeについて(テーマの利用＋個別のカスタマイズ)
- レジスタの活用法について
- LSPやスニペット、補完の活用法について

**Reply by makizo** [reply]
あと、情報入手先の紹介や更新、使い方、っていうのも有難いかもしれないです。
